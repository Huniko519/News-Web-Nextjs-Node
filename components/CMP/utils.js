export const isCCPAJurisdiction = () => global.PageCriteria.geo === 'US';
export const isGDPRJurisdiction = () => global.PageCriteria.inEU;

export const getCmpName = () => (isCCPAJurisdiction() ? '__uspapi' : '__cmp');
export const getCmp = () => global[getCmpName()];
// eslint-disable-next-line no-return-assign
export const setCmp = (cmp) => (global[getCmpName()] = cmp);
export const shouldConsentManagementBeEnforced = () => isCCPAJurisdiction() || isGDPRJurisdiction();

export const hasConsent = () => localStorage.getItem('mol.ads.cmp.consent') === 'yes';

const CCPA_COMMANDS_TO_CACHE = {
  getUSPData: '__cachedUSPData',
};

const GDPR_COMMANDS_TO_CACHE = {
  getConsentData: '__cachedGetConsentData',
  getVendorConsents: '__cachedVendorConsents',
};

const CACHED_CMP_COMMANDS = () => (
  isCCPAJurisdiction() ? CCPA_COMMANDS_TO_CACHE : GDPR_COMMANDS_TO_CACHE
);

/**
 * Function to cache certain CMP commands in localstorage
 */
export const updateCache = async () => {
  const cmp = getCmp();
  return Promise.all(Object.entries(CACHED_CMP_COMMANDS()).map(
    ([cmd, key]) => new Promise((resolve) => {
      cmp.a.push([cmd, null, (data, success) => {
        if (success) {
          localStorage.setItem(key, JSON.stringify(data));
          resolve();
        }
      }]);
    }),
  ));
};

/**
 * Function to create a provisional __cmp / __ccpa command queue
 */
export const createProvisionalCMP = async () => {
  // eslint-disable-next-line no-multi-assign
  const cmp = global.__cmpProvisional = getCmp() || ((cmd, param, cb) => {
    const key = CACHED_CMP_COMMANDS()[cmd];
    if (key && localStorage.getItem(key)) {
      cb(JSON.parse(localStorage.getItem(key)), true);
    } else {
      getCmp().a.push([cmd, param, cb]);
    }
  });

  setCmp(cmp);

  cmp.a = cmp.a || [];

  updateCache();
  cmp('onUpdate', null, (_, success) => success && updateCache());
};
