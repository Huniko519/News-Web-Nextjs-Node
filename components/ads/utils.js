import React from 'react';
import ampAdRTCConfig from './amp-ad-rtc';

/**
 * Builds the markup needed for the MPU placeholders
 *
 * @param string pos Ad position suffix i.e l1
 *
 * @return jsx
 */
export const buildMpuAd = (pos) => {
  if (!pos) {
    return null;
  }

  return (
    <>
      <div className="inews__advert" id={`mpu_${pos}`} data-pos={`mpu_${pos}`} />
      <div className="inews__advert" id={`mpu_tablet_${pos}`} data-pos={`mpu_tablet_${pos}`} />
      <div className="inews__advert" id={`mpu_mobile_${pos}`} data-pos={`mpu_mobile_${pos}`} />
    </>
  );
};

/**
 * Generates the full slot name for an amp-ad
 *
 * @param slot full Ad position identifier i.e mpu_amp_l1
 * @param object wp-json post object
 *
 * @return jsx
 */
export const generateAmpAdSlot = (slot, post) => {
  const categories = post.breadcrumbs.map((b) => b.slug);
  return `/5765/inews.amp/${categories.join('_')}/${slot}`;
};

/**
 * Generates the targeting json object for an amp-ad
 *
 * @param object wp-json post object
 *
 * @return json string
 */
export const generateAmpAdJson = (post) => {
  const json = {
    targeting: {
      article: post.id,
      rtc: 1,
    },
  };
  return JSON.stringify(json);
};

/**
 * Generates the RTC json object for an amp-ad
 *
 * @param object wp-json post object
 *
 * @return json string
 */
export const generateAmpAdRTC = (slot) => {
  const config = ampAdRTCConfig[slot];

  if (!config) {
    return {};
  }

  const rtc = {
    vendors: {
      criteo: {
        ZONE_ID: config.criteo ?? '',
        LINE_ITEM_RANGES: '0..10:0.01;10..25:0.05;25..50:0.10;50..100:0.25',
      },
      prebidappnexus: {
        PLACEMENT_ID: config.appnexus ?? '',
      },
    },
    urls: [
      `https://amp.casalemedia.com/amprtc?v=1&w=${config.primaryW || 300}&h=${config.primaryH || 250}&ow=ATTR(data-override-width)&oh=ATTR(data-override-height)&ms=ATTR(data-multi-size)&s=${config.indexexchange ?? ''}&p=CANONICAL_URL&consent_state=CONSENT_STATE&consent_string=CONSENT_STRING`,
      `https://aax.amazon-adsystem.com/e/dtb/bid?src=3065&pubid=PUB_UUID&amp=1&u=CANONICAL_URL&slots=%5B%7B%22sd%22%3A%22${slot}%22%2C%22s%22%3A%5B%22ATTR(width)xATTR(height)%22%5D%2C%22ms%22%3A%22ATTR(data-multi-size)%22%7D%5D&pj={"amp":1}&gdprc=CONSENT_STRING`,
    ],
  };

  return JSON.stringify(rtc);
};

/**
 * Builds the complete amp-ad element with all of the needed
 * attributes
 *
 * @param string pos Ad position suffix i.e l1
 * @param object wp-json post object
 *
 * @return jsx
 */
export const buildMpuAmpAd = (pos, post) => {
  if (!pos || !post?.id) {
    return null;
  }
  const adPos = `mpu_amp_${pos}`;
  const props = {
    'id': adPos,
    'layout': 'fluid',
    'height': 'fluid',
    'type': 'doubleclick',
    'json': generateAmpAdJson(post),
    'rtc-config': generateAmpAdRTC(adPos),
    'data-slot': generateAmpAdSlot(adPos, post),
    'data-multi-size': '300x250',
    'data-block-on-consent': '',
    'data-npa-on-unknown-consent': 'true',
  };

  if (adPos === 'mpu_amp_l1') {
    props['data-multi-size'] = '300x250,320x480';
    props['data-multi-size-validation'] = 'false';
  }

  return <div className="inews__advert inews__advert_amp"><amp-ad {...props} /></div>;
};

export const renderBillboardAd = (context) => {
  let render = true;

  if (context?.router?.pathname === '/category' && context?.sidebar?.['_embedded']?.widgets?.[0]) {
    render = false;
  }

  return render;
};
