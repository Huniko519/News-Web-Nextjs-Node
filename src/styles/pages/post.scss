@import '../partials/_colors';
@import '../partials/_dimensions';
@import '../partials/_media_queries';
@import '../partials/_fonts';
@import '../partials/_moreFrom';
@import '../partials/post-layout/post-items/post-items';

/* stylelint-disable-next-line selector-no-qualifying-type */
body.is-post {
  padding-top: calc(
    #{$sticky-archive-title-height}+#{$sticky-header-height-mobile} + 1rem
  );

  @media (--tablet-and-up) {
    padding-top: calc(
      #{$sticky-archive-title-height}+#{$sticky-header-height-tablet}+ 1rem
    );
  }

  @media (--desktop-and-up) {
    padding-top: calc(
      #{$sticky-archive-title-height}+#{$sticky-archive-title-height}+#{$sticky-header-height-desktop}+1rem
    );
  }
}

.inews__main__primary__wrapper {
  display: flex;
  flex: 0 1 auto;
  flex-direction: row;
  flex-wrap: wrap;
  box-sizing: border-box;
  margin-right: -0.5rem;
  margin-left: -0.5rem;
}

.inews__main__primary {
  position: relative;
  padding: 0.5rem 1.5rem 1rem;
  overflow: hidden;

  @media (--tablet-and-up) {
    padding: 0.5rem 1rem 1rem;
  }

  @media (--desktop-and-up) {
    flex: 0 0 auto;
    flex-basis: 66.66666667%;
    box-sizing: border-box;
  }

  .article-content {
    @include clearfix;
    font-family: $serif-copy;

    // Cropping and hiding post for users specified in a composer logic on Piano (see PostPianoLockContainer.js)
    &-cropped {
      max-height: 190px;
      overflow-y: hidden;
    }

    a {
      color: $red;
      transition: all 0.2s ease-in;

      &:hover {
        box-shadow: 0 -1px 0 0 $white inset, 0 -2px 0 0 $red inset;
        transition: all 0.2s ease-out;
      }
    }

    // Add the i logo to the very last paragraph in the article body
    > p:last-of-type {
      &::after {
        margin-left: 0.25rem;
        vertical-align: middle;
        content: url('/static/images/logo/i-post-logo.svg');
      }
    }

    ul {
      padding-bottom: 0.8rem;
      padding-left: 0;
      line-height: 1.3rem;
      list-style: none;
      border-bottom: 1px solid $grey-3;

      li {
        margin-bottom: 10px;

        &::before {
          display: inline-block;
          width: 10px;
          height: 10px;
          margin-right: 1rem;
          background-color: $red;
          content: '';
        }
      }
    }

    .inews__mpu {
      display: inline;
    }

    //clearing embeds to prevent distortion when next to a floated ad.
    amp-twitter,
    amp-facebook,
    amp-instagram,
    amp-iframe,
    figure {
      clear: both;
    }
  }

  .inews__email-signup {
    h3 span {
      display: inline;
    }

    @media (--mobile-and-below) {
      float: none;
      width: initial;
    }
  }

  .inews__advert {
    &:not(.teads) {
      float: right;
      margin-left: 20px;

      //@todo: Figure out how to add a label to AMP Ads nicely
      > div {
        position: relative;
        display: flex;
        justify-content: center;
        padding-top: 25px;
        padding-bottom: 1rem;

        &::before {
          position: absolute;
          top: 0;
          box-sizing: border-box;
          width: 300px;
          padding-left: 10px;
          color: #9b9b9b;
          font-size: 12px;
          line-height: 25px;
          background-color: #ececec;
          content: 'Advertisement';
        }
      }

      @media (--mobile-and-below) {
        float: none;
        margin: auto;
      }
    }
  }
}

.headline {
  margin: 1rem 0 0;
  font-weight: 900;
  font-size: 2.3rem;
  line-height: 2.5rem;
}

.excerpt {
  display: inline-block;
  margin: 1rem 0;
  font-weight: 400;
  font-size: 1.2rem;
  line-height: 1.6rem;
}

.inews__post-byline {
  margin: 0.5rem 0 1rem;
  background-color: var(--inews-color-secondary);

  &__avatar {
    width: 80px;
    height: 80px;
  }
}

.inews__post-byline__author-link {
  padding: 0.5rem;
  font-weight: 700;
  font-family: $serif-headline;
}

.inews__post-byline__author-job_title {
  margin-top: 0.5rem;
  color: $grey-2;
  font-weight: 400;
  font-size: 0.8rem;
  font-family: $sans;
}

.inews__post-share {
  amp-social-share {
    display: inline-block;
    width: 28px;
    height: 28px;
    margin: 0 2rem 0 0;
    background-color: $white;
    border-radius: 3px;
  }

  .amp-social-share-email {
    background-image: url('/static/images/social-share/email.svg');
    background-size: 170%;
  }

  .amp-social-share-twitter {
    background-image: url('/static/images/social-share/twitter.svg');
    background-size: 160%;
  }

  .amp-social-share-facebook {
    background-color: $grey-2;
  }

  .amp-social-share-whatsapp {
    background-color: $white;
    background-image: url('/static/images/social-share/whatsapp.svg');
    background-size: 110%;
  }
}

.inews__post-date {
  margin-bottom: 1rem;
  color: $grey-2;
  font-size: 0.8rem;
}

.inews__post__moddate {
  margin-left: 0.5rem;
}

.inews__disclaimer {
  font-size: 0.8rem;
  font-style: italic;
  line-height: 1.2rem;
}

.inews__post-byline__sponsored {
  &-logo {
    display: block;

    amp-img {
      width: 100%;
      max-width: 120px;
      height: auto;
      max-height: 120px;
    }
  }
}

.inews__topic-list {
  margin-bottom: 1rem;
  padding: 0.5rem 0;
  border-top: 1px solid $grey-2;
  border-bottom: 1px solid $grey-2;

  h4 {
    margin: 0 0 0.5rem;
    color: $grey-2;
    font-weight: 400;
    font-family: $serif-headline;
  }

  li {
    display: inline-block;
    color: $grey-2;

    &::after {
      padding-right: 0.25rem;
      content: '/';
    }

    &:last-child {
      &::after {
        content: '';
      }
    }

    a {
      padding: 0 0.25rem 0 0;
      color: $red;
      font-weight: 700;
      font-family: $serif-headline;
    }
  }
}

figure {
  figcaption {
    padding: 0.4rem 0;
    font-size: 0.8rem;
    font-family: $sans;
  }
}

.featured-media {
  position: relative;
  padding-top: 56.25%;
  overflow: hidden;

  > .featured-image,
  > .featured-video {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    max-width: 100%;
    height: 100%;
  }
}

.featured-image {
  figcaption {
    padding: 0.4rem 0.6rem 0.4rem 0;
  }
}

// This is just a hack for responsive 16:9 video for the DailyMotion player on web (not AMP)
.dailymotion-wrapper {
  position: relative;

  iframe {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
  }
}

blockquote {
  margin: 0;
  padding: 0;
  color: var(--inews-color-primary);
  font-size: 1.375rem;
  border-left: 0;
  /* stylelint-disable-next-line selector-no-qualifying-type */
  &.qa {
    display: block;
    clear: both;
    padding: 0.25rem 1rem;
    color: $black;
    font-size: 1rem;
    background: $grey-3;
    border-left: 0;
  }
}

.inews__shortcode-readmore {
  clear: both;
  min-height: 84px;
  margin: 1rem 0;
  background: $red;

  .inews__shortcode-readmore__text a {
    display: block;
    color: $white;
    line-height: 1.2rem;

    &:hover {
      box-shadow: none;
    }
  }
}

.inews__shortcode-readmore__image {
  float: left;
  width: 84px;
  margin-right: 0.5rem;
  background: $white;
}

.inews__shortcode-readmore__text {
  h4 {
    margin: 0;
    padding-top: 0.4rem;
    color: $white;
  }

  p {
    margin: 0;
    padding: 0.4rem 0 0.2rem;
    color: $white;
  }
}

.inews__shortcode-readmoreauto {
  margin: 0;
  padding: 0;
  background: transparent;
  border-top: 1px solid $grey-5;
  border-bottom: 1px solid $grey-5;

  h4 {
    margin: 0.5rem 0;
    text-transform: uppercase;
  }

  &_post {
    display: flex;
    margin: 0.75rem 1.5rem;
    color: $black;

    &:first-of-type {
      margin-top: 0.5rem;
    }

    &_image {
      amp-img {
        width: 87px;
        height: auto;
      }
    }

    &_text {
      display: flex;
      flex-direction: column;
      justify-content: space-between;
      padding: 0 0 0 14px;

      &_title {
        display: block;
        padding-bottom: 5px;
        font-weight: bold;
        font-family: $serif-headline;
        line-height: 1.3rem;
      }

      &_date {
        margin: 0;
        color: rgb(140, 140, 140);
        font-size: 11px;
        line-height: 11px;
      }
    }
  }
}

/* stylelint-disable selector-no-qualifying-type */
figure {
  &.wp-caption,
  &.wp-block-image {
    clear: both;
    width: 100%;
    max-width: 100%;
  }
}
/* stylelint-enable selector-no-qualifying-type */

.wp-block-columns {
  display: flex;
}

.inews__post__category-banner {
  position: static;
  top: 60px;
  left: 0;
  z-index: 999;
  box-sizing: border-box;
  width: 100%;
  height: 50px;
  overflow: hidden;
  color: var(--inews-color-primary);
  font-weight: 900;
  font-size: 1.5rem;
  font-family: $serif-headline;
  background-color: $white;
  border-bottom: 10px solid var(--inews-color-primary);

  @media (--tablet-and-up) {
    top: 150px;
    width: auto;
    margin: 0 -0.5rem;
  }

  @media (--desktop-and-up) {
    position: absolute;
    width: 100%;
    margin: 0;
    height: $sticky-archive-title-height;
  }

  &__parent-mobile {
    display: none;

    @media (--tablet-and-below) {
      display: inline;
    }
  }

  &__expend-mobile {
    position: absolute;
    display: none;

    a {
      color: $grey-7;
    }

    @media (--tablet-and-below) {
      display: block;
    }
  }

  @media (--tablet-and-below) {
    &.collapsed {
      max-height: 2.9rem;
      padding-right: 2rem;

      .inews__post__category-banner__expend-mobile {
        top: 1rem;
        right: 0.625rem;
      }
    }

    &.expended {
      .inews__post__category-banner__expend-mobile {
        right: 0.625rem;
        bottom: 1rem;
      }
    }
  }
}

.i__post__cat-ban__in {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  margin: 0 auto;
  padding: 0.5rem 1rem;

  @media (--desktop-and-up) {
    max-width: $main-content-width-desktop;
  }
}

.inews__post__category-parent-item {
  margin: 0 10px;
  @media (--mobileL-and-up) {
    margin: 0 1rem;
  }
}

.inews__post__category-item {
  margin: 0 0.625rem;
  padding: 7px 0;
  color: $grey-2;
  font-weight: normal;
  border-bottom: solid 1px transparent;
  @media (--mobileL-and-up) {
    margin: 0 1rem 0 0;
  }
}

.inews__post__category__grand-parent-item {
  color: var(--inews-color-primary);
  font-weight: bold;
  font-size: 1rem;
  font-family: $serif-headline;
}

.inews__post__category-item:hover {
  border-bottom-color: $grey-2;
}

@media (--tablet-and-below) {
  .inews__main__primary {
    max-width: 100%;
  }
}

@media (--mobile-only) {
  .inews__leaderboard {
    position: sticky;
    top: 50px;
    z-index: 100;
    margin-top: 0;
  }

  .inews__advert[data-pos='sticky_banner_top'] {
    position: relative;
    top: initial;
  }

  .headline {
    font-size: 1.5rem;
    line-height: 1.8rem;
  }

  .excerpt {
    font-size: 1rem;
    line-height: 1.4rem;
  }

  .inews__post-byline__author-link {
    font-size: 0.8rem;
  }

  .inews__post-byline__avatar {
    width: 60px;
    height: 60px;
  }

  .inews__post__moddate {
    margin-left: 5px;
  }

  .inews__topic-list {
    li {
      a {
        font-size: 0.8rem;
      }
    }
  }

  .inews__main > .row {
    margin-right: 0;
    margin-left: 0;
  }

  .inews__shortcode-readmore {
    margin: 1rem -1rem;

    .inews__shortcode-readmore__text {
      h4 {
        font-size: 0.8rem;
      }

      p {
        a {
          font-size: 0.9rem;
          line-height: 1.1rem;
        }
      }
    }
  }
}

// Push the DailyMotion mobile player under the top nav
#dailymotion-pip-small-viewport {
  /* stylelint-disable-next-line declaration-no-important */
  top: 116px !important;
  /* stylelint-disable-next-line declaration-no-important */
  z-index: 2 !important;
}

/* 
Container for loading a Piano template in order to crop and hide posts (see PostPianoLockContainer.js)
*/
.piano-lock-container {
  position: relative;
  margin-bottom: 1.5rem;

  &--show::before {
    position: absolute;
    right: 0;
    bottom: 100%;
    left: 0;
    width: 100%;
    height: 100px;
    background-image: linear-gradient(
      to top,
      rgba(255, 255, 255, 1) 0%,
      rgba(255, 255, 255, 1) 30%,
      rgba(255, 255, 255, 0) 100%
    );
    content: '';
  }
}
