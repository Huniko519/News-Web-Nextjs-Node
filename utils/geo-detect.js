const GEO_LOOKUP_TIMEOUT = 2000;

const dev = process.env.NODE_ENV === 'development';

const scriptBase = (dev || global.location?.host === 'inews.co.uk')
  ? 'inews.co.uk'
  : 'inewsint.co.uk';

export const scriptUrl = `//${scriptBase}/geo/locationjson-v1.3.html`;

/* eslint-disable no-async-promise-executor *//* @TODO: @rd13 to confirm eslint-disable */
export default () => new Promise(async (resolve) => {
  // Default / fallback geo incase lookup fails
  let PageCriteria = {
    geo: 'GB',
    region: '',
    city: '',
    inEU: true,
  };

  // Don't break anything if geo lookup fails
  try {
    const response = await Promise.race([
      fetch(scriptUrl),
      new Promise((reject) => setTimeout(() => reject(new Error('Request timeout')), GEO_LOOKUP_TIMEOUT)),
    ]);

    const {
      Country, REGION_CODE, CITY, INEU,
    } = await response.json();

    PageCriteria = {
      geo: Country,
      region: REGION_CODE,
      city: CITY,
      inEU: INEU,
    };
  } catch (error) {
    /* eslint-disable-next-line no-console */
    console.log('Geo lookup failed, using fallback geo.', error);
  }

  global.PageCriteria = PageCriteria;
  resolve(PageCriteria);
});
