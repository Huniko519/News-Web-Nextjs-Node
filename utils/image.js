export const getFeaturedImage = (post, size, fallbackImage) => {
  const image = post?._embedded?.['wp:featuredmedia']?.[0] ?? false;

  if ((!image || !image.id) && !fallbackImage) {
    return null;
  }

  const data = {
    src: fallbackImage,
    height: 0,
    width: 0,
    caption: '',
    alt: '',
  };

  if (image && image?.source_url) {
    data.src = image.source_url;
    data.width = image?.media_details?.width ?? 0;
    data.height = image?.media_details?.height ?? 0;
    data.alt = image?.alt_text ?? data.alt;
    data.caption = image?.caption?.rendered ?? data.caption;

    // Return custom size if available
    if (size && image?.media_details?.sizes?.[size]) {
      data.src = image.media_details.sizes[size].source_url;
      data.width = image.media_details.sizes[size].width;
      data.height = image.media_details.sizes[size].height;
    }
  }

  return data;
};

export const getCoAuthorImage = (coAuthor, size, fallbackImage, isOpinion = false) => {
  const image = coAuthor?.avatar_urls?.full ?? false;

  if (!image && !fallbackImage) {
    return null;
  }

  const data = {
    src: fallbackImage,
    height: 0,
    width: 0,
    caption: '',
    alt: '',
  };

  if (image) {
    data.src = image;
    data.width = 100;
    data.height = 100;
    data.alt = '';
    data.caption = '';

    if (coAuthor.display_name) {
      data.alt = `Avatar image for ${coAuthor.display_name}`;
    }

    // Return custom size if available
    if (size) {
      // Fetching author image for opinion template from opinion_avatar_urls if exists,
      // else size will be fetched from avatar_urls
      if (isOpinion && coAuthor.opinion_avatar_urls && coAuthor.opinion_avatar_urls[size]) {
        data.src = coAuthor.opinion_avatar_urls[size];
      } else if (coAuthor.avatar_urls[size]) {
        data.src = coAuthor.avatar_urls[size];
      }
    }
  }

  return data;
};
