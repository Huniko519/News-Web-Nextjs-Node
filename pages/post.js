import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { WP } from '../utils/wpapi';

import {
  isValidPostUrl, locatePost, hasPostBreadcrumb, getSidebarId, getMoreFromPosts,
} from '../utils/post';
import PostTemplate from '../components/post-templates/PostTemplate';

class Post extends Component {
  static async getInitialProps(context) {
    const post = await locatePost(context);

    if (post.error) {
      return { error: post.error };
    }

    // redirect to correct url for post
    if (!isValidPostUrl(context, post)) {
      context.res.writeHead(301, {
        Location: post.link,
      });
      context.res.end();
    }

    const sidebarId = getSidebarId(post);
    const moreFromPosts = await getMoreFromPosts(post);

    // Show opinion post sidebar only if post category === opinion and post has only one author
    const isOpinion = hasPostBreadcrumb(post, 'opinion');
    const numberOfAuthors = post['co-authors'] ? post['co-authors'].length : 0;
    let sidebar;
    if (isOpinion && numberOfAuthors === 1) {
      sidebar = await WP
        .sidebar()
        .id('opinion-post-sidebar')
        .param('coauthor', post['co-authors'][0].user_nicename)
        .embed()
        .then((data) => data)
        .catch(() => { });
    } else {
      sidebar = await WP
        .sidebar()
        .id(sidebarId)
        .embed()
        .then((data) => data)
        .catch(() => { });
    }
    return {
      post, sidebar, yoast_head: post.yoast_head, moreFromPosts,
    };
  }

  render() {
    return <PostTemplate {...this.props} />;
  }
}

Post.propTypes = {
  post: PropTypes.objectOf(PropTypes.any).isRequired,
  sidebar: PropTypes.objectOf(PropTypes.any),
  // eslint-disable-next-line react/forbid-prop-types
  moreFromPosts: PropTypes.array,
};

Post.defaultProps = {
  sidebar: {},
  moreFromPosts: [],
};

export const config = { amp: true };
export default Post;
