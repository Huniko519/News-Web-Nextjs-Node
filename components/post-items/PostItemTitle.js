import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import PageContext from '../PageContext';

import { getSocialTitle, getReviewRatingClassName } from '../../utils/post';

/**
 * Display the post title
 *
 * @param {*} props
 */
const PostItemTitle = (props) => {
  const { post } = props;
  const context = useContext(PageContext);

  let title = post.title.rendered;
  if (post.inews_homepage_headline && context && context?.router?.pathname === '/') {
    // On the homepage, use a homepage headline, if set
    title = post.inews_homepage_headline;
  } else if (post.inews_vertical_headline && context && context?.router?.pathname !== '/') {
    // See if the vertical headline is set
    title = post.inews_vertical_headline;
  } else if (post.inews_primary_headline) {
    // Use the primary headline if not on the homepage, and vertical headline not set
    title = post.inews_primary_headline;
  } else {
    // Fall back to the social title
    title = getSocialTitle(post);
  }

  return (
    <h2>
      <a
        className={getReviewRatingClassName(post)}
        href={post.link}
        title={post.title.raw}
        dangerouslySetInnerHTML={{ __html: title }}
      />
    </h2>
  );
};

/* istanbul ignore next */
PostItemTitle.propTypes = {
  post: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default PostItemTitle;
