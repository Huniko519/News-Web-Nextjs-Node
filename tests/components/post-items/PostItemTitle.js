import React from 'react';
import PostItemTitle from '../../../components/post-items/PostItemTitle';
import * as mockUtilPost from '../../../utils/post';

jest.mock('../../../utils/post', () => ({
  __esModule: true,
  getSocialTitle: jest.fn(() => 'Social Title'),
  getReviewRatingClassName: jest.fn(() => 'rating-class-name'),
}));

beforeEach(() => {
  jest.clearAllMocks();
});

test('renders the snapshot', () => {
  const post = {
    id: 1,
    title: {
      rendered: 'title',
    },
    breadcrumbs: [
      { name: 'News' },
    ],
    link: 'https://inews-article.com/test-article',
  };
  const postItemTitle = shallow(<PostItemTitle post={post} />);
  expect(postItemTitle).toMatchSnapshot();
});

it('should output a linked headline using the title returned from getSocialTitle', () => {
  const post = {
    id: 1,
    title: {
      rendered: 'title',
    },
    breadcrumbs: [
      { name: 'News' },
    ],
    link: 'https://inews-article.com/test-article',
  };
  const postItemTitle = shallow(<PostItemTitle post={post} />);
  expect(mockUtilPost.getSocialTitle).toHaveBeenCalled();
  expect(mockUtilPost.getReviewRatingClassName).toHaveBeenCalled();

  const header = postItemTitle.find('h2');
  expect(header.length).toBe(1);
  const anchor = header.find('a');
  expect(anchor.length).toBe(1);
  expect(anchor.prop('href')).toBe(post.link);
  expect(anchor.prop('title')).toBe(post.title.raw);
  expect(anchor.prop('className')).toBe('rating-class-name');
  expect(anchor.prop('dangerouslySetInnerHTML')).toMatchObject({
    __html: 'Social Title',
  });
});
