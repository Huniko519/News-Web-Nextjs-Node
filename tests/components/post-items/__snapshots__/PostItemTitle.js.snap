// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the snapshot 1`] = `
<h2>
  <a
    className="rating-class-name"
    dangerouslySetInnerHTML={
      Object {
        "__html": "Social Title",
      }
    }
    href="https://inews-article.com/test-article"
  />
</h2>
`;
